src/components/pagination/index.tsx

```diff
@@ -0,0 +1,259 @@
+import { useState, createContext, useContext, useMemo } from 'react'
+import {
+	type PropsWithChildren,
+	type Dispatch,
+	type SetStateAction,
+} from 'react'
+import classNames from 'classnames'
+
+import { IconChevronLeft16 } from '@hashicorp/flight-icons/svg-react/chevron-left-16'
+import { IconChevronRight16 } from '@hashicorp/flight-icons/svg-react/chevron-right-16'
+
+import s from './pagination.module.css'
+import { generateTruncatedList } from './helpers'
+import {
+	ButtonArrowProps,
+	ButtonNumberProps,
+	InfoProps,
+	NavProps,
+	PaginationProps,
+	SizeSelectorProps,
+} from './types'
+
+const PaginationContext = createContext({
+	totalItems: 0,
+	pageSize: 0,
+	setPageSize: (() => void 1) as Dispatch<SetStateAction<number>>,
+	page: 1,
+	setPage: (() => void 1) as Dispatch<SetStateAction<number>>,
+	totalPages: 0,
+})
+const usePagination = () => useContext(PaginationContext)
+
+const Pagination = ({
+	totalItems,
+	pageSize: _pageSize,
+	page: _page = 1,
+	onPageChange = () => void 1,
+	onPageSizeChange = () => void 1,
+	children,
+}: PropsWithChildren<PaginationProps>) => {
+	if (typeof totalItems !== 'number' || totalItems < 1) {
+		throw new Error(
+			'Pagination: totalItems is required, but was not specified. Please try passing a non-zero, positive value such as `103`.'
+		)
+	}
+	if (typeof _pageSize !== 'number' || _pageSize < 1) {
+		throw new Error(
+			'Pagination: pageSize is required, but was not specified. Please try passing a non-zero, positive value such as `10`.'
+		)
+	}
+	if (_page < 1) {
+		throw new Error(
+			'Pagination: page must be a non-zero, positive number. Please try passing a value such as `1`.'
+		)
+	}
+
+	const [page, _setPage] = useState(_page)
+	const [pageSize, _setPageSize] = useState(_pageSize)
+
+	const setPage = (val: number) => {
+		if (val !== page) {
+			_setPage(val)
+			onPageChange(val)
+		}
+	}
+
+	const setPageSize = (val: number) => {
+		if (val !== pageSize) {
+			_setPageSize(val)
+			onPageSizeChange(val)
+
+			// by default, updating page size will reset the page to "1"
+			setPage(1)
+		}
+	}
+
+	return (
+		<PaginationContext.Provider
+			value={{
+				totalItems,
+				pageSize,
+				setPageSize,
+				page,
+				setPage,
+				totalPages: Math.ceil(totalItems / pageSize),
+			}}
+		>
+			<div className={s.pagination}>{children}</div>
+		</PaginationContext.Provider>
+	)
+}
+
+const Info = ({ showTotalItems = true }: InfoProps) => {
+	const pagination = usePagination()
+	const start = pagination.page * pagination.pageSize - pagination.pageSize + 1
+	const end = Math.min(
+		pagination.page * pagination.pageSize,
+		pagination.totalItems
+	)
+	return (
+		<div className={s.info}>
+			{start} - {end}
+			{showTotalItems ? ` of ${pagination.totalItems}` : null}
+		</div>
+	)
+}
+
+const Nav = ({ type = 'compact' }: NavProps) => {
```

Comment:
Non-blocking question: Can we expose classNames option for the compound components so that we can use CSS to conditionally hide them vs needing JS media query hook?

Reply:
Yes, 100% — I was on the fence about this, but seems this is the cleaner approach

Reply:
I may actually punt on this for now — the `Nav.type` prop handles some conditional rendering via TS/JS, and the use of classNames doesn't yet handle that case :/